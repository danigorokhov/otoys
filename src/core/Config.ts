import { DocumentType } from './types/document';

type DocumentRemote = {
    type: 'remote';
    url: string;
};
type DocumentLocal = {
    type: 'local';
    path: string;
};
type DocumentInline = {
    type: 'inline';
    content: string;
    documentType: DocumentType;
};

type DocumentLoader =
    | DocumentRemote
    | DocumentLocal
    | DocumentInline
;

export type ConfigOptions = {
    documentLoader: DocumentLoader;
    output: string;
    pathWhitelist?: string;
};

export class Config {
    public documentLoader: DocumentLoader;
    public output: string;
    public pathWhitelist?: string;

    constructor(configOptions: ConfigOptions) {
        this.documentLoader = configOptions.documentLoader;
        this.output = configOptions.output;
        this.pathWhitelist = configOptions.pathWhitelist;
    }
}
