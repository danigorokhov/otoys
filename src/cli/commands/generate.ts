import { Config } from '../../core/Config';
import { generate as generateTypes } from '../../core';

import { resolve as resolveConfig, validate as validateConfig } from '../config';

type CliOptions = Record<string, unknown> & {
    config: string;
};

type Generate = (cliOptions: CliOptions) => void;

export const generate: Generate = async (cliOptions) => {
    // TODO think of pure cli running
    const userConfig = resolveConfig(cliOptions.config);
    const userConfigValidated = validateConfig(userConfig);
    const config = new Config(userConfigValidated);

    generateTypes(config);
};
