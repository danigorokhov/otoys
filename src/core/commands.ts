import { Config } from './Config';
import { Registry } from './Registry';
import { DocumentHandler } from './DocumentHandler';
import { TypesGeneratorHandler } from './TypesGeneratorHandler';

type Generate = (config: Config) => void;

export const generate: Generate = async config => {
    // TODO to pure config without Registry
    const registry = new Registry({ config });

    const documentHandler = new DocumentHandler(registry);
    const document = await documentHandler.load();

    const generator = new TypesGeneratorHandler(registry, document);
    const _ast = generator.generate();
};
