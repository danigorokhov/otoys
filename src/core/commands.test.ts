import { expect, describe, it } from '@jest/globals';
import { OpenAPIObject } from 'openapi3-ts/oas30';

import petstore3 from '../../examples/v3/petstore3.json';
import onePasswordEvents from '../../examples/v3/one-password-events.json';
import { Config } from './Config';
import { generate } from './commands';

// TODO more tests
describe('commands/generate', () => {
    it('should generate types for petstore3 API', async () => {
        const document = petstore3 as unknown as OpenAPIObject;
        const config = new Config({
            documentLoader: {
                type: 'inline',
                content: JSON.stringify(document),
                documentType: 'json',
            },
            output: {
                type: 'inline',
            },
        });

        expect(await generate(config)).toMatchSnapshot();
    });

    it('should generate types for one-password-events API', async () => {
        // https://api.apis.guru/v2/specs/1password.com/events/1.0.0/openapi.json
        const document = onePasswordEvents as OpenAPIObject;
        const config = new Config({
            documentLoader: {
                type: 'inline',
                content: JSON.stringify(document),
                documentType: 'json',
            },
            output: {
                type: 'inline',
            },
        });

        expect(await generate(config)).toMatchSnapshot();
    });
});
