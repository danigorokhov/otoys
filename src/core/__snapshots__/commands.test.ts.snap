// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/generate should generate types for one-password-events API 1`] = `
"export type Introspection = {
    Features?: string[];
    IssuedAt?: DateTimeRFC3339;
    UUID?: string;
};
export type Error = {
    Error?: {
        Message?: string;
    };
};
export type ItemUsageItems = {
    items?: ItemUsage[];
} & CursorCollection;
export type SignInAttemptItems = {
    items?: SignInAttempt[];
} & CursorCollection;
export type Cursor = {
    cursor?: string;
};
export type ResetCursor = {
    end_time?: DateTimeRFC3339;
    limit?: number;
    start_time?: DateTimeRFC3339;
};
export type DateTimeRFC3339 = string;
export type CursorCollection = Cursor & {
    has_more?: boolean;
};
export type ItemUsage = {
    action?: Action;
    client?: Client;
    item_uuid?: UUID;
    timestamp?: DateTimeRFC3339;
    used_version?: number;
    user?: User;
    uuid?: UUID;
    vault_uuid?: UUID;
};
export type SignInAttempt = {
    category?: "success" | "credentials_failed" | "mfa_failed" | "modern_version_failed" | "firewall_failed" | "firewall_reported_success";
    client?: Client;
    country?: string;
    details?: Details;
    session_uuid?: UUID;
    target_user?: User;
    timestamp?: DateTimeRFC3339;
    type?: "credentials_ok" | "mfa_ok" | "password_secret_bad" | "mfa_missing" | "totp_disabled" | "totp_bad" | "totp_timeout" | "u2f_disabled" | "u2f_bad" | "u2f_timout" | "duo_disabled" | "duo_bad" | "duo_timeout" | "duo_native_bad" | "platform_secret_disabled" | "platform_secret_bad" | "platform_secret_proxy" | "code_disabled" | "code_bad" | "code_timeout" | "ip_blocked" | "continent_blocked" | "country_blocked" | "anonymous_blocked" | "all_blocked" | "modern_version_missing" | "modern_version_old";
    uuid?: UUID;
};
export type Action = unknown;
export type Client = {
    app_name?: string;
    app_version?: string;
    ip_address?: string;
    os_name?: string;
    os_version?: string;
    platform_name?: string;
    platform_version?: string;
};
export type UUID = string;
export type User = {
    email?: string;
    name?: string;
    uuid?: UUID;
};
export type Details = {
    value?: string;
};
"
`;

exports[`commands/generate should generate types for petstore3 API 1`] = `
"export type Pet = {
    id?: number;
    name: string;
    category?: Category;
    photoUrls: string[];
    tags?: Tag[];
    status?: "available" | "pending" | "sold";
};
export type Order = {
    id?: number;
    petId?: number;
    quantity?: number;
    shipDate?: string;
    status?: "placed" | "approved" | "delivered";
    complete?: boolean;
};
export type User = {
    id?: number;
    username?: string;
    firstName?: string;
    lastName?: string;
    email?: string;
    password?: string;
    phone?: string;
    userStatus?: number;
};
export type ApiResponse = {
    code?: number;
    type?: string;
    message?: string;
};
export type Category = {
    id?: number;
    name?: string;
};
export type Tag = {
    id?: number;
    name?: string;
};
"
`;
